de.peeeq.wurstscript/src/main/java/de/peeeq/wurstscript/intermediatelang/optimizer/SideEffectAnalyzer.java

```diff
@@ -219,4 +289,14 @@ public boolean cannotUseVar(ImStmt s, ImVar v) {
             return directlyUsedVariables(s).contains(v);
         }
     }
+
+    /**
+     * Checks if the given function calls any functions or modifies and variable
+     */
+    public boolean hasSideEffects(ImFunctionCall call) {
+        Set<ImFunction> natives = calledNatives(call.getFunc());
+        Set<ImFunction> directFuncs = calledFunctions(call.getFunc());
+        Set<ImVar> imVars = directlySetVariables(call.getFunc());
+        return natives.size() + directFuncs.size() + imVars.size() > 0;
+    }
```

Comment:
1. You could define this function to take an arbitrary Element like the other functions, making it more general.
2. You could write this as `!calledNatives(call.getFunc()).isEmpty() && !calledFunctions(call.getFunc()).isEmpty() && !directlySetVariables(call.getFunc()).isEmpty()` that would make the computation a bit faster in some cases.